dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,655 @@
+    this.verifyNotClosed();
+    this.closed = true;
+    return this._flush();
+  }
+
+  private async _flush(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    const writePromises = trackedBatches.map(batch => batch.awaitBatch());
+    this.sendBatches();
+    await Promise.all(writePromises);
+  }
+
+  private verifyNotClosed(): void {
+    if (this.closed) {
+      throw new Error('BulkWriter has already been closed.');
+    }
+  }
+
+  /**
+   * Gradually ramps up the maximum number of operations per second by 50%
+   * every 5 minutes in line with the 500/50/5 rule.
+   */
+  private rampMaxOps(): void {
+    delayExecution(() => {
+      this.maxOpsPerSecond = this.maxOpsPerSecond * 1.5;
+      this.rampMaxOps();
+    }, 5 * 60 * 1000);
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = this.batchQueue.length > 0;
+    for (const bulkBatch of this.batchQueue) {
+      if (!bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+      if (bulkBatch.canAddDoc(ref)) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        console.warn(
+          '[BulkWriter]',
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
+    } else {
+      return batch;
+    }
+  }
+
+  private createNewBatch(): BulkCommitBatch {
+    const newBatch = new BulkCommitBatch(
+      this.getId(),
+      this.firestore.batch(),
+      this.maxBatchSize
+    );
+
+    this.batchQueue.push(newBatch);
+    return newBatch;
+  }
+
+  private getId(): number {
+    const id = this.currentId;
+    if (this.currentId === Number.MAX_SAFE_INTEGER) {
+      this.currentId = 0;
+    } else {
+      this.currentId++;
+    }
+    return id;
+  }
+
+  /**
+   * Attempts to send batches starting from the front of the BatchQueue until a
+   * batch cannot be sent.
+   *
+   * After a batch is complete, try sending batches again.
+   */
+  private sendBatches(): void {
+    const unsentBatches = this.batchQueue.filter(
+      batch => batch.sendTime === null
+    );
+
+    let index = 0;
+    while (
+      unsentBatches.length > index &&
+      this.isBatchSendable(unsentBatches[index])
+    ) {
+      const batch = unsentBatches[index];
+      batch.refsInBatch.forEach(batch => {
+        this.refsInFlight.add(batch);
+      });
+
+      batch.bulkCommit().then(results => {
+        batch.resolveResults(results);
+
+        const batchIndex = this.batchQueue.indexOf(batch);
+        assert(batchIndex !== -1, 'The batch should be in the BatchQueue');
+        this.batchQueue.splice(batchIndex, 1);
```

Comment:
I would prefer if we simplified this logic to send all batches up until the point where we cannot send the next batch. We are already logging a warning - that should be enough to discourage this usage pattern.